### PR TITLE
BIG-15160: Updates to client library

### PR DESCRIPTION
**Pull request by aleachjr**
ping @lord2800 @splittingred @philipmuir @lovek323

### CODE REVIEW

File: src/Bigcommerce/Api/Client.php [hunk 1]
```diff
@@ -9,900 +9,1014 @@
...
-	}
-
-
-
+    static private $store_url;
+    static private $username;
+    static private $api_key;
+    static private $connection;
+    static private $resource;
+    static private $path_prefix = '/api/v2';
```
Comment (lovek323) [inline src/Bigcommerce/Api/Client.php]: Any chance for docblocks on these?

File: src/Bigcommerce/Api/Client.php [hunk 2]
```diff
@@ -9,900 +9,1014 @@
...
+            throw new Exception("'store_url' must be provided");
+        }
+
+        if (!isset($settings['username'])) {
+            throw new Exception("'username' must be provided");
+        }
+
+        if (!isset($settings['api_key'])) {
+            throw new Exception("'api_key' must be provided");
+        }
```
Comment (lovek323) [inline src/Bigcommerce/Api/Client.php]: :+1: This is a good way to handle this.

File: src/Bigcommerce/Api/Client.php [hunk 3]
```diff
@@ -9,900 +9,1014 @@
...
+     * Requires a settings array to be passed in with the following keys:
+     *
+     * - store_url
+     * - username
+     * - api_key
+     *
+     * @param array $settings
+     * @throws \Exception
+     */
+    public static function configure($settings)
```
Comment (lovek323) [inline src/Bigcommerce/Api/Client.php]: Add typehint `array $settings`?

File: src/Bigcommerce/Api/Client.php [hunk 4]
```diff
@@ -9,900 +9,1014 @@
...
+     */
+    public static function useJson()
+    {
+        self::connection()->useXml(false);
+    }
+
+    /**
+     * Switch SSL certificate verification on requests.
+     */
+    public static function verifyPeer($option = false)
```
Comment (lovek323) [inline src/Bigcommerce/Api/Client.php]: Any chance of having this default to `true` like the other ones? I just think `$api->verifyPeer()` would suggest that I am asking for the peer to be verified.

File: src/Bigcommerce/Api/Client.php [hunk 5]
```diff
@@ -9,900 +9,1014 @@
...
+    {
+        self::connection()->verifyPeer($option);
+    }
+
+    /**
+     * Set which cipher to use during SSL requests.
+     */
+    public static function setCipher($cipher = 'rsa_rc4_128_sha')
```
Comment (lovek323) [inline src/Bigcommerce/Api/Client.php]: I wonder if we could create this as a constant, e.g., `DEFAULT_SSL_CIPHER` or something.

File: src/Bigcommerce/Api/Client.php [hunk 6]
```diff
@@ -9,900 +9,1014 @@
...
+    static private $connection;
+    static private $resource;
+    static private $path_prefix = '/api/v2';
+
+    /**
+     * Full URL path to the configured store API.
+     *
+     * @var string
+     */
+    static public $api_path;
```
Comment (lovek323) [inline src/Bigcommerce/Api/Client.php]: Is it possible to make this camelCase?

File: src/Bigcommerce/Api/Connection.php [hunk 7]
```diff
@@ -8,455 +8,469 @@
...
-		if (!$this->responseStatusLine && strpos($headers, 'HTTP/') === 0) {
-			$this->responseStatusLine = $headers;
-		} else {
-			$parts = explode(': ', $headers);
-			if (isset($parts[1])) {
-				$this->responseHeaders[$parts[0]] = trim($parts[1]);
-			}
-		}
+    /**
+     * @var cURL resource
```
Comment (lovek323) [inline src/Bigcommerce/Api/Connection.php]: If this is of type `resource`, can you put that in, e.g.,

```php
/**
 * @var resource cURL resource
 */
```

File: src/Bigcommerce/Api/Connection.php [hunk 8]
```diff
@@ -8,455 +8,469 @@
...
+     */
+    private $curl;
+
+    /**
+     * @var hash of HTTP request headers
```
Comment (lovek323) [inline src/Bigcommerce/Api/Connection.php]: Can we make this

```php
/**
 * @var array Hash of HTTP request headers.
 */
```

Comment (lovek323) [inline src/Bigcommerce/Api/Connection.php]: There are some other docblocks below that have missing types as well.

File: src/Bigcommerce/Api/Connection.php [hunk 9]
```diff
@@ -8,455 +8,469 @@
...
+
+    /**
+     * Number of redirects followed in a loop.
+     * @var int
+     */
+    private $redirectsFollowed = 0;
+
+    /**
+     * Deal with failed requests if failOnError is not set.
+     * @var mixed
```
Comment (lovek323) [inline src/Bigcommerce/Api/Connection.php]: Instead of `mixed`, could we say `string | false` or something like that? (I know `false` doesn't seem like a type, but it is what phpdoc suggests.)

File: src/Bigcommerce/Api/Connection.php [hunk 10]
```diff
@@ -8,455 +8,469 @@
...
+        curl_setopt($this->curl, CURLOPT_PUT, false);
+        curl_setopt($this->curl, CURLOPT_HTTPGET, false);
+        curl_setopt($this->curl, CURLOPT_POSTFIELDS, $body);
+        curl_exec($this->curl);
+
+        return $this->handleResponse();
+    }
+
+    /**
+     * Make an HTTP HEAD request to the specified endpoint.
```
Comment (lovek323) [inline src/Bigcommerce/Api/Connection.php]: Can we have a `@return` please?

Comment (lovek323) [inline src/Bigcommerce/Api/Connection.php]: There are several other instances of this. It's not a hard requirement for merge, but it would be great to generate documentation for this using phpdoc.

File: src/Bigcommerce/Api/Client.php (hunk 4 above)
Comment (lord2800) [inline src/Bigcommerce/Api/Client.php]: It actually needs to be an int (0/1/2, with 2 being the behavior most people expect when they think of verifying a peer).

**Comment by lovek323** [comment]
Apart from very minor nitpicks, this LGTM :+1:

File: src/Bigcommerce/Api/Client.php (hunk 1 above)
Comment (aleachjr) [inline src/Bigcommerce/Api/Client.php]: sure, why not :smile:

File: src/Bigcommerce/Api/Client.php (hunk 4 above)
Comment (lord2800) [inline src/Bigcommerce/Api/Client.php]: In fact, we can perhaps just map the true/false to 2 (verify cert and hostname) or 0 (don't verify).

Comment (aleachjr) [inline src/Bigcommerce/Api/Client.php]: The only potential issue here is we are changing existing behavior. While this is going to be a new major release version, it still may be confusing.

**Comment by philipmuir** [comment]
we should default a new tls cipher, we don't support RC4 on our production servers anymore from memory. Or at least a note in the readme that we don't recommend using it

**Comment by aleachjr** [comment]
Confirmed that removing default cipher connect I am still able to connect on my VM, integration, and prod store.

**Comment by aleachjr** [comment]
ping @chrisboulton

File: test/Unit/Api/ClientTest.php [hunk 11]
```diff
@@ -305,14 +305,14 @@ public function resources()
     {
         return array(
             //     path              function          classname
-            array('products',       '%sProduct',      'Product'),
```
Comment (lord2800) [inline test/Unit/Api/ClientTest.php]: Could you undo this change? These were aligned purposely.

File: test/Unit/Api/ClientTest.php [hunk 12]
```diff
@@ -251,17 +251,17 @@ public function collections()
     {
         return array(
             //      path           function             classname
-            array('products',      'getProducts',      'Product'),
```
Comment (lord2800) [inline test/Unit/Api/ClientTest.php]: Same here.

File: test/Unit/Api/ClientTest.php [hunk 13]
```diff
@@ -92,20 +92,20 @@ public function testVerifyPeerPassesThroughToConnection()
...
         $this->connection->expects($this->exactly(2))
-             ->method('setCipher')
-             ->withConsecutive(
-                 array('rsa'),
-                 array('rsa_rc4_128_sha') // this is the default value
-             );
+            ->method('setCipher')
+            ->withConsecutive(
+                array('rsa'),
+                array('tls') // this is the default value
```
Comment (lord2800) [inline test/Unit/Api/ClientTest.php]: The point of this test was to make it use the default value specified in the function arguments, rather than specifying it directly.

**Comment by lord2800** [comment]
:+1:

**Comment by xenph** [comment]
:+1: For the removal of the mandatory cipher.
